lib/descriptor/descriptor.js

```diff
@@ -0,0 +1,196 @@
+/*!
+ * descriptor.js - descriptor object for bcoin
+ * Copyright (c) 2023, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const {stringType, addChecksum} = require('./common');
+
+/**
+ * Descriptor
+ * Base class for all descriptors.
+ * Represents an output script
+ * @alias module:descriptor.Descriptor
+ * @property {String} name name of descriptor function
+ * @property {PubkeyProvider[]} pubkeys public key arguments for the descriptor
+ * (size 1 for PK, PKH, WPKH; any size for WSH and Multisig).
+ * @property {Descriptor[]} subdescriptors  sub-descriptor arguments
+ * for the descriptor (empty for everything but SH and WSH)
+ * @property {Network} network
+ */
+
+class Descriptor {
+  /**
+   * Create a descriptor
+   * @constructor
+   */
+
+  constructor() {
+    this.name = null;
+    this.pubkeys = [];
+    this.subdescriptors = [];
+    this.network = null;
+  }
+
+  /**
+   * Test whether the descriptor contains a private key.
+   * @returns {Boolean}
+   */
+
+  hasPrivateKeys() {
+    for (const key of this.pubkeys) {
+      if (key.getPrivatekey()) {
+        return true;
+      }
+    }
+    for (const subdesc of this.subdescriptors) {
+      if (subdesc.hasPrivateKeys()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Test whether the descriptor contains public/private keys
+   * in the form of HD chains
+   * @returns {Boolean}
+   */
+
+  isRange() {
+    for (const key of this.pubkeys) {
+      if (key.isRange()) {
+        return true;
+      }
+    }
+    for (const subdesc of this.subdescriptors) {
+      if (subdesc.isRange()) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Whether this descriptor has all information about signing
+   * (ignoring private keys).
+   * Returns false only for `addr` and `raw` type.
+   * @returns {Boolean}
+   */
+
+  isSolvable() {
+    for (const subdesc of this.subdescriptors) {
+      if (!subdesc.isSolvable()) {
+        return false;
+      }
+    }
+    return true;
```

Comment:
Why are we only checking sub-descriptors here?
If the descriptor is `addr` or `raw`, we should return `false` directly.

Reply:
```suggestion
    return this.subdescriptors.every((desc) => desc.isSolvable());
```

Reply:
addr and raw descriptors overrides isSolvable function by returning false.